### ISSUE TITLE
Shutdown issue with detached actors

### ISSUE THREAD

**Issue by Neverlord** [actor-framework/actor-framework#1196]
We've observed occasional deadlocks with detached actors where CAF would block on `await_detached_threads` during shutdown. The actor system reaches this point only after `await_all_actors_done` returned, which might point to a bug in `caf::detail::private_thread`. When spawning all actors without the `detached` flag, we could no longer produce any shutdown issue.

**Comment by mavam** [comment]
This might be related to #1110.

**Comment by Neverlord** [comment]
Could be. That issue talked about issues with detached actors in combination with the registry, though. In our case we didn't use the registry. Maybe we're lucky and the underlying cause is the same.

**Comment by mavam** [comment]
> Maybe we're lucky and the underlying cause is the same.

Exactly my hope. 🤞

**Comment by mavam** [comment]
I ported the example to work with master and tried it out. The hang still persists. Then I checked out `issue/11961` after the possibly related fix. The issue is gone. I get a valid termination, closing this issue now.

For the record, here is the adapted example:

```cpp
#include <caf/all.hpp>

using namespace caf;

behavior aut(event_based_actor* self) {
  return {
    [](std::string x) {
      return x;
    }
  };
}

void caf_main(actor_system& sys) {
  sys.registry().put("aut", sys.spawn<detached>(aut));
  auto x = sys.registry().get("aut");
  anon_send_exit(x, exit_reason::user_shutdown);
  // Comment to experience the hang.
  sys.registry().erase("aut");
}

CAF_MAIN()
```

**Comment by sourcedelica** [comment]
Is it possible to backport this fix to 0.17?

**Comment by Neverlord** [comment]
> Is it possible to backport this fix to 0.17?

Probably, but we rather focus our limited resources on the current release version. Especially since there's a workaround for earlier versions (not using `detached`). That being said, we have no policy against backports. You might give it a shot yourself if this is important to you and file a PR against `release/0.17`.

**Comment by sourcedelica** [comment]
Ok - I'll give it a shot and see if it's feasible for me to do it.
